src/main.rs

```diff
@@ -50,13 +50,65 @@ use std::os::unix::fs::PermissionsExt;
 use std::os::unix::net::{UnixListener, UnixStream};
 use std::path::Path;
 use std::thread;
+use std::time::Duration;
 
 use anyhow::{Context, Result};
 use slog::{debug, error, o, Drain};
 
 mod ffi;
 mod handlers;
 mod protocol;
+mod semaphore;
+
+const SOCKET_PATH: &str = "/var/run/nscd/socket";
+const MAX_THREADS: usize = 64;
+const MAX_THREADS_WAIT_TIMEOUT: Duration = Duration::from_secs(5);
+
+fn main() -> Result<()> {
+    ffi::disable_internal_nscd();
+
+    let decorator = slog_term::TermDecorator::new().build();
+    let drain = slog_term::FullFormat::new(decorator).build().fuse();
+    let drain = slog_async::Async::new(drain).build().fuse();
+
+    let logger = slog::Logger::root(drain, slog::o!());
+
+    let path = Path::new(SOCKET_PATH);
+    std::fs::create_dir_all(path.parent().expect("socket path has no parent"))?;
+    std::fs::remove_file(path).ok();
+    let listener = UnixListener::bind(path).context("could not bind to socket")?;
+    std::fs::set_permissions(path, std::fs::Permissions::from_mode(0o777))?;
+
+    let sem = semaphore::Semaphore::new(MAX_THREADS);
+
+    for stream in listener.incoming() {
+        match stream {
+            Ok(stream) => {
+                let thread_logger = logger.clone();
+
+                // block on acquring a permit before spawning the thread so
+                // that we actually limit the number of threads used.
+                let permit = match sem.acquire(MAX_THREADS_WAIT_TIMEOUT) {
+                    Ok(p) => p,
+                    Err(_) => {
+                        anyhow::bail!("timed out waiting to spawn a handler thread. blowing up!")
```

Comment:
@leifwalsh @geofft this should be a better error message. Thoughts on what it should be?

Reply:
Is there a reason you went with limiting the number of threads with a semaphore, but still forking on each request, instead of preforking a threadpool of that many threads and bailing if we can't get one of them to accept our request within the timeout?

Reply:
I remember having a conversation about it in slack, but can't remember why we decided on this. I can put up a PR with a [threadpool](https://crates.io/crates/threadpool) tonight so we can compare.